### DISCUSSION TITLE
How can I send contents of a file to OpenAI?

### DISCUSSION THREAD

**Discussion by nandgator** [microsoft/teams-sdk#337] [Q&A] [answered]
I'm able to acquire a file uploaded by the user with `attachments.downloadUrl` and save it to a file say `transcript.txt`, but how do I send that to OpenAI? OpenAI doesn't seem to pick up uploaded file contents directly.

Here's my `skprompt.txt`

```text
The following is a conversation with an AI assistant, called Epitoma.
Epitoma is an expert in text summarization and Human is the individual interacting with Epitoma.
Epitoma should always understand and summarize the text input messages sent by the Human.
Epitoma should always summarize the contents of the transcript files (transcript.txt) uploaded by the Human.

{{$history}}
Human: {{$input}}
Epitoma:
```

Let's say I've added such activity:

```ts
import { readFileSync } from 'fs';

...

// Listen for ANY message to be received.
// MUST BE AFTER EVERY OTHER MESSAGE HANDLERS
app.activity(
  ActivityTypes.Message,
  async (context: TurnContext, state: ApplicationTurnState) => {
    const sampleFile = readFileSync('./transcript.txt', 'utf-8');
    // How do send contents of sampleFile to OpenAI?
  }
);
```

**Comment by aacebo** [comment]
did you configure your app with the ai settings? You should be able to specify the prompt folder and name. It will read the file from disk for you but it would need to be named `skprompt.txt` and have a `config.json` to accompany it.

If you want to send the file to the planner directly try using the apps `planner` interface or extend the `OpenAIPlanner` and pass it in.

**Reply by nandgator** [reply]
> did you configure your app with the ai settings? You should be able to specify the prompt folder and name. It will read the file from disk for you but it would need to be named skprompt.txt and have a config.json to accompany it.

Well all that is indeed in place.

![image](https://github.com/microsoft/teams-ai/assets/38415384/49c590b4-d1b4-455a-b405-2a8235a78c90)

What I cannot figure out is how to send the contents of a text file (uploaded by bot user) to OpenAI?

> If you want to send the file to the planner directly try using the apps `planner` interface or extend the `OpenAIPlanner` and pass it in.

Is "the file" referring to `skprompt.txt`, `config.json` or user uploaded file?

**Reply by aacebo** [reply]
do you have a link to the source I can use to take a more detailed look? Or can you paste your entire `index.ts` file here without any credentials or sensitive information?

**Reply by Stevenic** [reply, marked as answer]
@joe733 you can register a function with the prompt manager that will do this:

```
ai.prompts.addFunction('getInput', async (context, state) => {
    return readFileSync('./transcript.txt', 'utf-8');
});
```

Then your prompt template would be:

```
The following is a conversation with an AI assistant, called Epitoma.
Epitoma is an expert in text summarization and Human is the individual interacting with Epitoma.
Epitoma should always understand and summarize the text input messages sent by the Human.
Epitoma should always summarize the contents of the transcript files (transcript.txt) uploaded by the Human.

{{$history}}
Human: {{getInput}}
Epitoma:
```

**Reply by nandgator** [reply]
I've registered a function but, AI does not seem to responds as expected. [Source](https://gitea.com/joe733/epitoma/src/commit/41b070e3efa9155d0775f53e7910839160689bb3/src/epitomaBot.ts#L131-L139).

```ts
// Create attachments bot that will handle incoming messages
const atmBot = new AttachmentsBot();

app.ai.prompts.addFunction(
  "Summarize",
  async (context: TurnContext, state: ApplicationTurnState) => {
    atmBot.run(context);
    return readFileSync("./downloads/transcript.txt", "utf-8");
  }
);
```

![Screenshot from 2023-08-22 11-00-56](https://github.com/microsoft/teams-ai/assets/38415384/e0cecd0c-9358-41f1-a57b-486d4d8562c5)

**Reply by aacebo** [reply]
@joe733 the `Summarize` function needs to be called from the prompt via `{{Summarize}}`, I don't see that happening in your source code.

**Reply by nandgator** [reply]
> @joe733 the `Summarize` function needs to be called from the prompt via `{{Summarize}}`, I don't see that happening in your source code.

@aacebo can you please point in the [source code](https://gitea.com/joe733/epitoma/src/commit/41b070e3efa9155d0775f53e7910839160689bb3/src/epitomaBot.ts) where am I missing this?

Also, I couldn't find this suggestion in the documentation. It would really help if you could point to that as well.

**Reply by aacebo** [reply]
@joe733 prompt functions get called from the prompt template https://gitea.com/joe733/epitoma/src/commit/41b070e3efa9155d0775f53e7910839160689bb3/src/prompts/chat/skprompt.txt which are documented here a bit https://github.com/microsoft/teams-ai/blob/main/getting-started/01.TEMPLATES.md

**Reply by aacebo** [reply]
If you look up at steves example, he defines the function `getInput` which is referenced in his template via `{{getInput}}`

**Reply by nandgator** [reply]
I followed [`PROMPT-TEMPLATES.md`](https://github.com/microsoft/teams-ai/blob/efc459952575cdf64c2859e2fc94c3dbde25f6b9/getting-started/01.PROMPT-TEMPLATES.md). [Changes](https://gitea.com/joe733/epitoma/commit/455365bbd66235549fa3bb0d8540b06b539b1c76).


```text
The following is a conversation with an AI assistant, called Epitoma.
Epitoma is an expert in text summarization and Human is the individual interacting with Epitoma.
Epitoma should always understand and summarize the text input messages sent by the Human.
Epitoma should always summarize the contents of the transcript files (transcript.txt) uploaded by the Human.

{{$history}}
Human: {{$summarize}}
Epitoma:
```

```ts
// const atmBot = new AttachmentsBot();

app.ai.prompts.addFunction(
  "summarize",
  async (context: TurnContext, state: ApplicationTurnState) => {
    // atmBot.run(context);
    const transcript = readFileSync("./downloads/transcript.txt", "utf-8");
    console.log(transcript);
    return transcript;
  }
);
```

```console
$ s -la src/downloads 
Permissions Size User  Group Date Modified Name
.rw-r--r--  2.2k us-er us-er 20 Sep 10:28  transcript.txt
```


... but the AI is as baffled as I am.


| ![image](https://github.com/microsoft/teams-ai/assets/38415384/830e421d-14cf-4c2c-8dd7-46bb9e985115) | ![image](https://github.com/microsoft/teams-ai/assets/38415384/4717d4e9-7a59-44a6-aa3e-0ccd842c3004) |
| --- | --- |

**Reply by aacebo** [reply]
@joe733 `Human: {{$summarize}}` should be `Human: {{summarize}}` since you are try to use a function and not a variable.

**Reply by nandgator** [reply]
Ah my bad 🤦‍♂️, how did I miss it?

Thank you @aacebo & @Stevenic!
